topics/server-api-key-auth.md

```diff
@@ -0,0 +1,283 @@
+                ?.let { AppPrincipal(it) }
+        }
+    }
+}
+```
+
+#### Customize key location {id="key-location"}
+
+By default, the `apiKey` provider looks for the API key in the `X-API-Key` header.
+
+You can use `headerName` to specify a custom header:
+
+```kotlin
+apiKey("api-key-header") {
+    headerName = "X-Secret-Key"
+    validate { key ->
+        // Validate and return principal
+    }
+}
+```
+
+### Step 2: Validate API keys {id="validate"}
+
+The validation logic depends on your application's requirements. Here are common approaches:
+
+#### Static key comparison {id="static-key"}
+
+For simple cases, you can compare against a predefined key:
+
+```kotlin
+apiKey {
+    validate { keyFromHeader ->
+        val expectedApiKey = environment.config.property("api.key").getString()
+        keyFromHeader
+            .takeIf { it == expectedApiKey }
+            ?.let { AppPrincipal(it) }
+    }
+}
+```
+
+> Store sensitive API keys in configuration files or environment variables, not in source code.
+>
+{style="note"}
+
+#### Database lookup {id="database-lookup"}
+
+For multiple API keys, validate against a database:
+
+```kotlin
+apiKey {
+    validate { keyFromHeader ->
+        // Look up the key in database
+        val user = database.findUserByApiKey(keyFromHeader)
+        user?.let { UserIdPrincipal(it.username) }
+    }
+}
+```
+
+#### Multiple validation criteria {id="multiple-criteria"}
+
+You can implement complex validation logic:
+
+```kotlin
+apiKey {
+    validate { keyFromHeader ->
+        val apiKey = database.findApiKey(keyFromHeader)
+
+        // Check if key exists, is active, and not expired
+        if (apiKey != null &&
+            apiKey.isActive &&
+            apiKey.expiresAt > Clock.System.now()
+        ) {
+            UserIdPrincipal(apiKey.userId)
+        } else {
+            null
+        }
+    }
+}
+```
+
+### Step 3: Configure challenge {id="challenge"}
+
+You can customize the response sent when authentication fails using the `challenge` function:
+
+```kotlin
+apiKey {
+    validate { key ->
+        // Validation logic
+    }
+    challenge { defaultScheme, realm ->
+        call.respond(
+            HttpStatusCode.Unauthorized,
+            "Invalid or missing API key"
+        )
+    }
+}
+```
+
+### Step 4: Protect specific resources {id="authenticate-route"}
+
+After configuring the `apiKey` provider, you can protect specific resources in your application using the
+[`authenticate`](server-auth.md#authenticate-route) function. In the case of successful authentication, you can
+retrieve an authenticated principal inside a route handler using the `call.principal` function.
+
+```kotlin
+routing {
+    authenticate {
+        get("/") {
+            val principal = call.principal<AppPrincipal>()!!
+            call.respondText("Hello, authenticated client! Your key: ${principal.key}")
+        }
+    }
+}
+```
+
+## Complete example {id="complete-example"}
```

Comment:
Wouldn't it be more descriptive if the heading is called "API Key authentication example" or "API Key authentication complete example" ?